src/execution.jl

```diff
@@ -299,26 +318,39 @@ function execute(dsl_file::String,
                  sub_comm_size::Int=1,
                  max_amplitudes::Union{Int, Nothing}=nothing,
                  max_parameters::Union{Int, Nothing}=nothing)
-
+    # Parse the dsl file to create a list of commands to execute in a context. Also parse
+    # the parameter file to get partition parameters and to create a sampler object to 
+    # produce bitstrings to compute amplitudes for.
     commands = parse_dsl(dsl_file)
+    sampler_args, partition_params = parse_parameters(param_file,
+                                                      max_amplitudes=max_amplitudes, 
+                                                      max_parameters=max_parameters)
 
-    bitstrings, partition_params = parse_parameters(param_file,
-                                                    max_amplitudes=max_amplitudes, max_parameters=max_parameters)
-
+    # Create a context to execute the commands in.
     ctx = QXContext(commands, partition_params, input_file)
     if comm !== nothing
         ctx = QXMPIContext(ctx, comm, sub_comm_size)
+
+        # Variables needed by a sampler to divide work evenly amongst groups of nodes.
+        sampler_args[:params][:rank] = MPI.Comm_rank(comm) ÷ sub_comm_size
+        sampler_args[:params][:comm_size] = MPI.Comm_size(comm) ÷ sub_comm_size
```

Comment:
This makes sense and I tried to change the sampler constructor to use multiple dispatch but I introduced some circular dependency between the sampling and execute modules and wasn't sure how to best get rid of it. 

I was trying to keep the constructor with the sampler types, so the Sampling module needs to know the Context types now, but the execute module which defines the context types was `using` the Sampling module.

I'm calling the sampler constructor using `QXContexts.Sampling.create_sampler` for the moment but would prefer to avoid this: https://github.com/JuliaQX/QXContexts.jl/pull/21/files#r628120261